File: 3.wuchongfuzifudezuichangzichuan.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastOccurrence = new HashMap<>();
        int left = 0;         
        int maxLen = 0;        

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);

            if (lastOccurrence.containsKey(ch) && lastOccurrence.get(ch) >= left) {
                left = lastOccurrence.get(ch) + 1;
            }

            lastOccurrence.put(ch, right);

            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: daima/3.wuchongfuzifudezuichangzichuan.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const lastOccurrence = new Map();
    let left = 0;
    let maxLen = 0;

    for (let right = 0; right < s.length; right++) {
        const ch = s[right];
        if (lastOccurrence.has(ch) && lastOccurrence.get(ch) >= left) {
            left = lastOccurrence.get(ch) + 1;
        }
        lastOccurrence.set(ch, right);
        maxLen = Math.max(maxLen, right - left + 1);
    }
    return maxLen;
};